src/main/scala/loamstream/googlecloud/HailCtlDataProcClient.scala

```diff
@@ -0,0 +1,94 @@
+package loamstream.googlecloud
+
+import java.nio.file.Paths
+
+import scala.util.Success
+import scala.util.Try
+
+import HailCtlDataProcClient.DelegateFn
+import loamstream.util.BashScript
+import loamstream.util.ExitCodes
+import loamstream.util.Loggable
+import loamstream.util.Tries
+
+/**
+ * @author clint
+ * Nov 28, 2016
+ */
+final class HailCtlDataProcClient private[googlecloud] (
+  googleConfig: GoogleCloudConfig,
+  hailConfig: HailConfig,
+  delegate: DataProcClient.CanStopAndTellIfRunning,
+  runCommand: DelegateFn = HailCtlDataProcClient.runCommand) extends DataProcClient with Loggable {
+
+  override def stopCluster(): Unit = delegate.stopCluster()
+
+  override def isClusterRunning: Boolean = delegate.isClusterRunning
+
+  override def startCluster(): Unit = {
+    debug(s"Starting cluster '${googleConfig.clusterId}'")
+
+    import HailCtlDataProcClient.startClusterTokens
+    
+    val exitCode = runCommand(googleConfig, hailConfig, startClusterTokens(googleConfig).mkString(" "))
+
+    ExitCodes.throwIfFailure(exitCode)
+  }
+}
+
+object HailCtlDataProcClient extends Loggable {
+  type DelegateFn = (GoogleCloudConfig, HailConfig, String) => Int
+  
+  def fromConfigs(googleConfig: GoogleCloudConfig, hailConfig: HailConfig): Try[HailCtlDataProcClient] = {
+    for {
+      wrapper <- CloudSdkDataProcWrapper.fromConfig(googleConfig)
+    } yield {
+      new HailCtlDataProcClient(googleConfig, hailConfig, new CloudSdkDataProcWrapper(googleConfig))
+    }
+  }
+
+  private[googlecloud] def startClusterTokens(config: GoogleCloudConfig): Seq[String] = {
+    val tokens: Seq[String] = Seq(
```

Comment:
zomg... This would likely be my first submit pass as well. Here's hoping in the future these tokens can all be done another way since I'm sure many of these are common among all commands (e.g. `--project` and `--zone`) and typos suck.

```scala
val projectToken = Seq("--project", projectId)
val zoneToken = Seq("--zone", zone)
val clusterToken = Seq(clusterId)
// ...

def token(args: Seq[Seq[String]], rest: String*) = args.flatten ++ rest
```

Reply:
There are definitely ways to improve this.  In `CloudSdkDataProcWrapper`, there's `gcloudTokens` which consolidates prepending common tokens for `gcloud` commands, like
```
gcloud +: "beta" +: "dataproc" +: "clusters" +: verb +: "--project" +: config.projectId
```
`hailctl` has a different CLI than `gcloud` though, so that method can't be used for `hailctl` commands.  Some _flags_ happen to be the same between both of those tools, like `--project`, but the cost/benefit of factoring out code for just that flag didn't seem high enough.